### PR TITLE
[Stacked PR 5/6] Add granular GDN remat policy and 8k/64k latency/memory benchmark suite

### PR DESCRIPTION
**Pull request by Rohan-Bierneni**
# Description

This is **PR 5 of 6** (the capstone PR #5098) in the stacked series establishing the canonical Pallas Gated Delta Net (GDN) implementation in MaxText.

This PR introduces:
1. **Granular Native GDN Rematerialization Policy**:
   - Adds native remat toggles `gdn: RematLocation` and `gdn_conv: RematLocation` (`'remat'`, `'device'`, `'offload'`) to `base.yml`, `types.py`, and `maxtext_utils.py`.
   - Adds cross-field configuration validation (`validate_gdn_remat_requires_kernel`) enforcing that setting `gdn` or `gdn_conv` to `'device'` or `'offload'` requires `use_gdn_kernel=True`. This prevents silent misconfiguration where Pure JAX would otherwise ignore the setting and redundantly recompute activations.
   - Adds `_expand_gdn_remat_names` in `maxtext_utils.py` expanding `"gdn"` to individual named residuals: `["gdn_core_attn_out", "gdn_chunk_states", "gdn_t_inv", "gdn_qkv", "gdn_b", "gdn_a"]` and `"gdn_conv"` to `["gdn_conv_out", "gdn_fwd_conv"]`.
   - Tags core GDN outputs and recurrence residuals with `jax.ad_checkpoint.checkpoint_name` in `gdn_bwd_pallas.py` and `qwen3.py`.
   - Implements `_get_gdn_aware_remat_policy` in `qwen3.py` ensuring that when custom remat is enabled (`tensors_on_device: ["decoder_layer_input", "gdn"]`), forward recurrence residuals are preserved in device memory, completely eliminating redundant forward kernel recomputation (~17 ms) during backward autodiff while keeping all Feed-Forward/MoE layers 100% rematerialized.
   - Adds unit tests in `gdn_bwd_pallas_test.py` and `pyconfig_test.py` covering gradient invariance, residual preservation, and configuration validation.

2. **Cloud TPU Latency & Memory Benchmark Suite (`tests/unit/gdn_benchmark_test.py`)**:
   - Provides standalone 8k and 64k Cloud TPU profiling scripts comparing Pure JAX against the Decoupled GDN Pallas kernel.
   - Measures isolated Pallas hardware calls (forward and backward) as well as the full monolithic decoder layer.

### 64k Ghostfish Cloud TPU Hardware Performance (1,024 Chunks, S=65,536, B=1, H_k=16, H_v=64, D=128):
- **Isolated Pallas Hardware Kernel Calls**:
  - Core Pallas Forward Kernel: **37.55 ms**
  - Core Pallas Backward Kernel: **93.59 ms**
- **Full Monolithic 1-Chip Decoder Layer (Projections + Gating + Conv1D + Kernels)**:
  - Forward Pass: **68.18 ms**
  - Backward Pass: **185.99 ms** (includes ~150 ms linear GEMM backprop on a single chip)
  - Total Step Latency: **254.17 ms**
  - Peak HBM Memory: **32.83 GB** (vs Pure JAX OOM)

# Files Changed
- `src/maxtext/configs/base.yml`: Added `gdn: 'remat'` and `gdn_conv: 'remat'` custom remat options.
- `src/maxtext/configs/types.py`: Added `gdn` and `gdn_conv` fields in `RematAndOffload`, included in custom remat lists, and added cross-field validator requiring `use_gdn_kernel=True`.
- `src/maxtext/utils/maxtext_utils.py`: Added `_expand_gdn_remat_names` and integrated with `get_save_and_offload_names`.
- `src/maxtext/models/kernels/gdn/gdn_bwd_pallas.py`: Tagged output and residuals with `checkpoint_name`.
- `src/maxtext/models/qwen3.py`: Added `_GDN_SAVED_NAMES`, `_get_gdn_aware_remat_policy`, and tagged `gdn_core_attn_out` and `out_proj`.
- `tests/unit/gdn_bwd_pallas_test.py`: Added `test_gdn_custom_remat_policy_preserves_residuals` and `test_gdn_granular_remat_requires_gdn_kernel`.
- `tests/unit/pyconfig_test.py`: Added `test_gdn_granular_remat_requires_gdn_kernel` and `test_gdn_granular_remat_accepts_gdn_kernel`.
- `tests/unit/gdn_benchmark_test.py`: Standalone 8k and 64k Cloud TPU latency and memory benchmark suite with isolated kernel timings.

# Tests
- Verified with `pre-commit run --files ...` (all hooks: pylint, pyink, codespell, yamllint passed cleanly).
- Google3 unit tests passed with `-c opt`: `//third_party/py/maxtext/tests/unit:gdn_bwd_pallas_test`, `//third_party/py/maxtext/tests/unit:pyconfig_test`, and `//third_party/py/maxtext/tests/unit:gdn_benchmark_ghostfish_test`.

# Checklist
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests and verified pre-commit linters pass.

### CODE REVIEW

File: src/maxtext/models/kernels/gdn/gdn_bwd_pallas.py [hunk 1]
```diff
@@ -0,0 +1,1636 @@
...
+  d_q_proj = jnp.sum(
+      d_q_rep.reshape(chunk_size, num_kq_heads, repeats, kq_head_dim), axis=2
+  )
+  d_k_proj = jnp.sum(
+      d_k_rep.reshape(chunk_size, num_kq_heads, repeats, kq_head_dim), axis=2
+  )
+
+  if use_qk_norm_in_gdn:
+    d_q_scaled = d_q_proj * scale
+    r_q = jnp.sqrt(jnp.sum(q_orig**2, axis=-1, keepdims=True) + 1e-12)
```
[the remark this answers is not in the document]
Comment (Rohan-Bierneni) [reply src/maxtext/models/kernels/gdn/gdn_bwd_pallas.py]: Resolved. In PR 3 and PR 5, epsilon for QK L2 normalization is standardized to `1e-6` in both the forward kernel and backward state computations.

File: src/maxtext/kernels/megablox/pallas_mosaic_tpu_v2_tgmm_kernel.py [hunk 2]
```diff
@@ -221,6 +221,9 @@ def make_tgmm_configs(
   size_lhs_sublane = min(size_lhs_sublane, size_m)
   size_rhs_sublane = pltpu.get_tpu_info().get_sublane_tiling(rhs.dtype)
   size_rhs_sublane = min(size_rhs_sublane, size_m)
+  common_sublane = min(size_lhs_sublane, size_rhs_sublane)
+  size_lhs_sublane = common_sublane
+  size_rhs_sublane = common_sublane
```
[the remark this answers is not in the document]
Comment (Rohan-Bierneni) [reply src/maxtext/kernels/megablox/pallas_mosaic_tpu_v2_tgmm_kernel.py]: Removed gmmv2 fixes out of these stacked prs

File: src/maxtext/models/kernels/gdn/gdn_bwd_pallas.py [hunk 3]
```diff
@@ -0,0 +1,1636 @@
...
+    q_unit = q_orig / r_q
+    d_q = (
+        d_q_scaled
+        - q_unit * jnp.sum(d_q_scaled * q_unit, axis=-1, keepdims=True)
+    ) / r_q
+
+    r_k = jnp.sqrt(jnp.sum(k_orig**2, axis=-1, keepdims=True) + 1e-12)
```
[the remark this answers is not in the document]
Comment (Rohan-Bierneni) [reply src/maxtext/models/kernels/gdn/gdn_bwd_pallas.py]: Resolved. In PR 3 and PR 5, epsilon for QK L2 normalization is standardized to 1e-6 in both the forward kernel and backward state computations.
